webapp/javascript/services/render.ts

```diff
@@ -126,6 +127,123 @@ export async function mergeWithQueryID(
   return Result.err(parsed.error);
 }
 
+const HeatmapSchema = z.object({
+  startTime: z.number(),
+  endTime: z.number(),
+  minValue: z.number(),
+  maxValue: z.number(),
+  minDepth: z.number(),
+  maxDepth: z.number(),
+  timeBuckets: z.number(),
+  valueBuckets: z.number(),
+  values: z.array(z.array(z.number())),
+});
+
+export interface getHeatmapProps {
+  query: string;
+  from: string;
+  until: string;
+  minValue: number;
+  maxValue: number;
+  heatmapTimeBuckets: number;
+  heatmapValueBuckets: number;
+  maxNodes?: string;
+}
+
+export type Heatmap = z.infer<typeof HeatmapSchema>;
+export interface HeatmapOutput {
+  heatmap: Heatmap;
+  profile: Profile;
+}
+
+export async function getHeatmap(
+  props: getHeatmapProps,
+  controller?: {
+    signal?: AbortSignal;
+  }
+): Promise<Result<HeatmapOutput, RequestError | ZodError>> {
+  // todo use common builder
```

Comment:
If you're going to leave this comment here can you elaborate on what you mean by this?